### PR TITLE
feat: add return_probs option to expose per-detection class probabili…

### PR DESCRIPTION
**Pull request by woreom**
…ties

- Add return_probs: bool = False config option (predict mode).
- non_max_suppression gains return_probs parameter; when True it returns (output, probs_out) where probs_out[i] is a (num_boxes, nc) tensor of raw class probability scores tracked through the same confidence/class/ NMS filters as the box tensor.
- Boxes gains an optional class_probs kwarg and a class_probs property exposing the (N, nc) tensor, or None when return_probs was not used.
- Results.__init__ now accepts a pre-built Boxes object directly to avoid double-wrapping.
- DetectionPredictor.postprocess, construct_results, and construct_result are updated to propagate class_probs end-to-end.
- Test: test_predict_return_probs verifies shape (N, nc) and None default.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a new `return_probs` prediction option to expose per-detection class probability tensors through `Results.boxes.class_probs` for YOLO detection outputs. 🎯

### 📊 Key Changes
- Added `return_probs: False` to `ultralytics/cfg/default.yaml` as a new prediction-time config option.
- Extended detection postprocessing in `ultralytics/models/yolo/detect/predict.py` to pass `return_probs` into NMS and carry per-image class probability tensors into final `Results` objects.
- Updated `ultralytics/utils/nms.py` so `non_max_suppression()` can optionally return per-detection class probability matrices alongside filtered detections.
- Enhanced `ultralytics/engine/results.py`:
- `Results` now preserves prebuilt `Boxes` objects.
- `Boxes` accepts optional `class_probs` data.
- Added a `Boxes.class_probs` property for downstream access.
- Added a Python test covering both behaviors:
- `return_probs=True` attaches `(N, nc)` class probability tensors.
- `return_probs=False` keeps `class_probs` as `None`. ✅

### 🎯 Purpose & Impact
- Enables users to inspect full per-class confidence distributions for each retained detection, not just the top predicted class. 📦
- Improves support for advanced postprocessing, analysis, and custom application logic built on detection outputs.
- Keeps default behavior unchanged, so existing prediction workflows should remain fully backward compatible. 🔒
- Adds test coverage to reduce regression risk for this new detection feature.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @woreom, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you soon.

To ensure a seamless integration of your work, please review the following checklist:
-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/results.py [hunk 1]
```diff
@@ -877,6 +882,7 @@ def __init__(self, boxes: torch.Tensor | np.ndarray, orig_shape: tuple[int, int]
         super().__init__(boxes, orig_shape)
         self.is_track = n == 7
         self.orig_shape = orig_shape
+        self._class_probs = class_probs
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/results.py:943]: ⚠️ **HIGH**: `Boxes` now stores `_class_probs`, but the inherited `BaseTensor` helpers rebuild `Boxes` from `self.data` only. That means common operations like `results[0].cpu()`, `results[0].numpy()`, `boxes[i]`, or `.to(...)` will silently drop `class_probs`. Since this feature is exposed on `Boxes`, it should survive the standard tensor/device/slicing transformations as well.

File: ultralytics/utils/nms.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ def non_max_suppression(
     rotated: bool = False,
     end2end: bool = False,
     return_idxs: bool = False,
+    return_probs: bool = False,
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/nms.py:29]: ⚠️ **HIGH**: `return_probs` is only handled in the main NMS path below. The existing early `prediction.shape[-1] == 6 or end2end` return path still returns just `output`, so calling `model.predict(..., return_probs=True)` on an end-to-end detector will hand `postprocess()` a plain list and `preds, probs_list = preds` will fail. Please either propagate probabilities through that branch as well or explicitly disable `return_probs` when `end2end` is active.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The feature is wired cleanly through predict → NMS → Results, but there are two functional gaps to address before merge: `return_probs` currently breaks on the end-to-end NMS fast path, and `Boxes.class_probs` is not preserved across standard `Boxes`/`Results` transformations.

💬 Posted 2 inline comments

**Comment by woreom** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
Thanks for this! It overlaps #24614 ("Expose logits in Results"), which is approved and reuses the existing NMS `return_idxs` plumbing rather than adding a parallel `return_probs` return path — so it covers the same need (per-detection per-class scores) with a smaller surface. Closing this in favor of #24614. Thanks for the contribution.
